### ISSUE TITLE
Error cross-compiling for aarch64

### ISSUE THREAD

**Issue by erlanger** [SWI-Prolog/swipl-devel#382]
I am getting the following error when cross-compiling for Android **aarch64**:

```
-- Build files have been written to: /home/builder/.termux-build/swi-prolog/build
[194/830] Building C object src/CMakeFiles/libswipl.dir/pl-wam.c.o
FAILED: src/CMakeFiles/libswipl.dir/pl-wam.c.o
/home/builder/.termux-build/_cache/18-aarch64-21-v1/bin/clang --target=aarch64-none-linux-android --gcc-toolchain=/home/builder/.termux-build/_cache/18-aarch64-21-v1 --sysroot=/home/builder/.termux-build/_cache/18-aarch64-21-v1/sysroot -Dlibswipl_EXPORTS -Isrc -I/home
/builder/.termux-build/swi-prolog/src/src -I/data/data/com.termux/files/usr/include -Oz -I/data/data/com.termux/files/usr/include -Os -DNDEBUG -fPIC   -Wall -MD -MT src/CMakeFiles/libswipl.dir/pl-wam.c.o -MF src/CMakeFiles/libswipl.dir/pl-wam.c.o.d -o src/CMakeFiles/l
ibswipl.dir/pl-wam.c.o   -c /home/builder/.termux-build/swi-prolog/src/src/pl-wam.c
In file included from /home/builder/.termux-build/swi-prolog/src/src/pl-wam.c:225:
/home/builder/.termux-build/swi-prolog/src/src/pl-alloc.c:1044:2: error: "Unsupported int64_t alignment conversion"
#error "Unsupported int64_t alignment conversion"
 ^
1 error generated.
[201/830] Building C object src/CMakeFiles/libswipl.dir/pl-comp.c.o
ninja: build stopped: subcommand failed.
```

Any idea why this could be happening?

**Comment by erlanger** [comment]
Strangely I get the same error if I cross-compile for Android **x86_64**:

```
-- Build files have been written to: /home/builder/.termux-build/swi-prolog/build
[141/830] Building C object src/CMakeFiles/libswipl.dir/pl-wam.c.o
FAILED: src/CMakeFiles/libswipl.dir/pl-wam.c.o
/home/builder/.termux-build/_cache/18-x86_64-21-v1/bin/clang --target=x86_64-none-linux-android --gcc-toolchain=/home/builder/.termux-build/_cache/18-x86_64-21-v1 --sysroot=/home/builder/.termux-build/_cache/18-x86_64-21-v1/sysroot -Dlibswipl_EXPORTS -Isrc -I/home/bui
lder/.termux-build/swi-prolog/src/src -I/data/data/com.termux/files/usr/include -Oz -I/data/data/com.termux/files/usr/include -Os -DNDEBUG -fPIC   -Wall -MD -MT src/CMakeFiles/libswipl.dir/pl-wam.c.o -MF src/CMakeFiles/libswipl.dir/pl-wam.c.o.d -o src/CMakeFiles/libsw
ipl.dir/pl-wam.c.o   -c /home/builder/.termux-build/swi-prolog/src/src/pl-wam.c
In file included from /home/builder/.termux-build/swi-prolog/src/src/pl-wam.c:225:
/home/builder/.termux-build/swi-prolog/src/src/pl-alloc.c:1044:2: error: "Unsupported int64_t alignment conversion"
#error "Unsupported int64_t alignment conversion"
 ^
1 error generated.
[148/830] Building C object src/CMakeFiles/libswipl.dir/pl-comp.c.o
ninja: build stopped: subcommand failed.
```

**Comment by erlanger** [comment]
Found out what the problem is. I am using a [TryRunResults.cmake](https://github.com/erlanger/termux-packages/blob/swi-prolog/packages/swi-prolog/TryRunResults.cmake) file to cache alignment values because we don't have a usable emulator at cross-compiling time to run `cmake/AlignOf.cmake
`. The `TryRunResults.cmake` file has values that work on 32bit arches only.

Do you think there is any way we can figure out the alignment values at compile time in the same way we're figuring out the endianness?

**Comment by JanWielemaker** [comment]
I doubt it.  Endianness is a binary representation, so you just create a binary with some magic value and look for a pattern in the generated binary.  Alignment restrictions are not so easy to find.  Possibly you can create a global struct with -say- a char and an int64_t and see how many bytes padding will appear in the structure.   Not that alignment has two aspects: strictly needed, meaning that if you do not obey you get a bus error or preferred, meaning unaligned access is safe but slow (i.e., x86, x64)

I think (but this needs some checking) that SWI-Prolog always tries to keep values aligned to their size (which is typically best for speed), but there are a couple of places where this is not possible and we either need to insert padding or need to copy the data to an aligned place before accessing it.  This
mostly concerns doubles and int64_t on 32-bit hardware.

Note that modern gcc has `alignof`, which may be used as a fallback for cross-compiling.  Using the safe `sizeof` will work too.  Finally we could add a preprocessor switch on the CPU as this is primarily a CPU property (I think AIX fixed bad aligned access over an OS trap but that is so slow that you want to avoid that.   Guess that should suffice to find a way out.  The value isn't terribly important.  Alignments smaller than the object size that are well supported by the CPU provide some code reduction and speedup, but barely measurable.

**Comment by erlanger** [comment]
Okay, I think I found a solution similar to the endianness trick, *and* it works at compile time.  `C++11` supports the `alignof` keyword otherwise we  fall back to `sizeof`. This code will produce a pattern "INT64_T_ALIGNMENT" in the binary which we can pattern match to get the alignment.

```c++
#include <cstdint>

int main() {
   static const unsigned char int64_t_alignment[18] = {
         'I', 'N','T','6','4','_','T','_','A','L','I','G','N','M','E','N','T',
#if __cplusplus >= 201103L
         alignof(int64_t)
#else
         sizeof(int64_t)
#endif
   };

   const int SIZE=sizeof(int64_t_alignment);
   unsigned char *p, *d, i;

   //Prevent optimizer from eliminating the code
   unsigned char dummy[SIZE];
   d = dummy;
   p = (unsigned char*) int64_t_alignment;
   for (i = 0; i < SIZE; ++i) {
     *d++ = *p++;
   }
   return dummy[SIZE-1]; //won't be used
}
```
What do you think?

I will extend it to `double` and whatever else we need if you think it's ok.

**Comment by JanWielemaker** [comment]
Cute! (in Dutch that is generally positive :)  It might be a bit of an overkill, but if it works I'm happy to accept the patch.  Surely others will like it too :)

**Comment by erlanger** [comment]
> Cute! (in Dutch that is generally positive :)

Hehehe. I had a big laugh with that one. In some languages it's more like an insult :)
PR being pushed now :)

**Comment by erlanger** [comment]
Closing, fixed by PR #383.
